### PR TITLE
Import votes from JSON

### PR DESCRIPTION
**Pull request by Ked57**
see #38

### CODE REVIEW

File: functions/src/send-email-to-manager.ts [hunk 1]
```diff
@@ -24,6 +24,10 @@ export const sendEmailToManager = functions.firestore
     }
 
     const vote = voteSnapshot.data()! as Vote;
+    if (!("voteFromUi" in vote)) {
```
Comment (hgwood) [inline functions/src/send-email-to-manager.ts]: That's an unconventional way to put it. Why not `!vote.voteFromUi`?

File: functions/src/import-votes.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    try {
+      validVotes = fromJSONtoVote(passedJSONString);
+    } catch (e) {
+      console.error(e);
+      res.sendStatus(500);
+    }
+
+    for (const validVote of validVotes) {
+      await db.collection("vote").add(validVote);
+    }
```
Comment (hgwood) [inline functions/src/import-votes.ts]: Use the batch feature to insert all the votes in one go :) https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes

File: functions/src/import-votes.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    const passedData = req.get("Authorization") || "";
+    const passedKey =
+      passedData.toString() === `Bearer ${importVotesConfigs.key}`;
+    if (!passedKey) {
+      console.info("Passed the wrong auth key, aborting");
+      return;
+    }
+    const passedJSONString = req.body.votesData;
+    const date: firebase.firestore.Timestamp = firebase.firestore.Timestamp.fromDate(
```
Comment (hgwood) [inline functions/src/import-votes.ts]: No need to spell out the type here, it can be inferred.

File: functions/src/import-votes.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+export const importVotes = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    if (!importVotesConfigs.enabled) {
+      console.info("Feature import votes disabled, aborting");
+      return;
+    }
+    const passedData = req.get("Authorization") || "";
+    const passedKey =
+      passedData.toString() === `Bearer ${importVotesConfigs.key}`;
```
Comment (hgwood) [inline functions/src/import-votes.ts]: No need for `toString`

File: functions/src/import-votes.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+export const importVotes = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    if (!importVotesConfigs.enabled) {
+      console.info("Feature import votes disabled, aborting");
+      return;
+    }
+    const passedData = req.get("Authorization") || "";
+    const passedKey =
```
Comment (hgwood) [inline functions/src/import-votes.ts]: I think you can find better names for those 2 vars. :)

File: functions/src/import-votes.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  const validVotes: Vote[] = [];
+  for (const supposedlyValidVote of supposedlyValidVotes.votes) {
+    if (
+      "value" in supposedlyValidVote &&
+      "campaign" in supposedlyValidVote &&
+      "agency" in supposedlyValidVote
+    ) {
+      validVotes.push(supposedlyValidVote);
+    }
+  }
```
Comment (hgwood) [inline functions/src/import-votes.ts]: That's a filter :) use `filter` for better clarity

File: functions/src/import-votes.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+const db = firebase.firestore();
+
+const fromJSONtoVote = (JSONString: string) => {
+  const supposedlyValidVotes = JSON.parse(JSONString);
+  const validVotes: Vote[] = [];
+  for (const supposedlyValidVote of supposedlyValidVotes.votes) {
+    if (
+      "value" in supposedlyValidVote &&
+      "campaign" in supposedlyValidVote &&
+      "agency" in supposedlyValidVote
```
Comment (hgwood) [inline functions/src/import-votes.ts]: That's correct but not really idiomatic. This what I would do:

```
for (const { value, campaign, agency } of supposedlyValidVotes.votes) {
  if (value && campaign && agency) ...
}
```

File: functions/src/import-votes.ts (hunk 6 above)
Comment (Ked57) [reply functions/src/import-votes.ts]: filter is way cooler indeed 😄

File: functions/src/import-votes.ts [hunk 8]
```diff
@@ -29,28 +25,25 @@ export const importVotes = functions.https.onRequest(
       console.info("Feature import votes disabled, aborting");
       return;
     }
-    const passedData = req.get("Authorization") || "";
-    const passedKey =
-      passedData.toString() === `Bearer ${importVotesConfigs.key}`;
-    if (!passedKey) {
+    const passedAuthData = req.get("Authorization") || "";
+    const passedAuthKey = passedAuthData === `Bearer ${importVotesConfigs.key}`;
```
Comment (hgwood) [inline functions/src/import-votes.ts]: Use the same names as in #35. :)

File: functions/src/import-votes.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  async (req: functions.Request, res: functions.Response) => {
+    if (!importVotesConfigs.enabled) {
+      console.info("Feature import votes disabled, aborting");
+      return;
+    }
+    const passedAuthData = req.get("Authorization") || "";
+    const passedAuthKey = passedAuthData === `Bearer ${importVotesConfigs.key}`;
+    if (!passedAuthKey) {
+      console.info("Passed the wrong auth key, aborting");
+      return;
```
Comment (hgwood) [inline functions/src/import-votes.ts]: This won't trigger any error and the function will timeout, use res code 401 instead

File: functions/src/import-votes.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      console.error(e);
+      res.sendStatus(500);
+    }
+    const voteBatch = db.batch();
+    for (const validVote of validVotes) {
+      voteBatch.create(db.collection("vote").doc(), validVote);
+    }
+    voteBatch
+      .commit()
+      .then(() => res.sendStatus(200), () => res.sendStatus(500));
```
Comment (hgwood) [inline functions/src/import-votes.ts]: use .catch instead

File: functions/src/index.ts [hunk 11]
```diff
@@ -22,6 +22,11 @@ export { getCampaign } from "./get-campaign";
 
 export { castVote } from "./cast-vote";
 
+
+export { updateStats } from "./update-stats";
```
Comment (hgwood) [inline functions/src/index.ts]: I think this should not be there

Comment (Ked57) [reply functions/src/index.ts]: We split updateStats on another PR, i don't know why it's in this one but it should be here

**Pull request merged**